tools/dart_tools/bin/build_jsoncpp.dart

```diff
@@ -0,0 +1,77 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the 'License');
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an 'AS IS' BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:io';
+
+import 'package:args/args.dart';
+
+Future<void> main(List<String> arguments) async {
+  if (!Platform.isWindows) {
+    throw new Exception('Building jsoncpp libraries is only available on '
+        'windows.');
+  }
+
+  final parser = new ArgParser()
+    ..addFlag('debug', abbr: 'd', negatable: false);
+  final args = parser.parse(arguments);
+
+  final downloadDirectory = arguments[0];
+  final debug = args['debug'];
+
+  await Process.run('git', [
+    'init',
```

Comment:
Why the `init/remote add/fetch` chain instead of just checking whether the folder is already present and then cloning?

Reply:
Because GN creates the parent folders before running a script, including subdirectories, of known output files and git won't clone into a non-empty folder. I didn't want to delete the GN created folders because the script would re-clone the repo for every build. With this chain it will fetch only if necessary.
If we are moving the clone to another directory then that would be the correct way to do it.

Reply:
Sigh. The interdependent hacks of the jsoncpp/GN interaction is getting pretty messy.

Maybe the right way to proceed incrementally here is:

- Write the script to clone, rather than have it contort to fit GN's needs
- Have the Windows GN instructions, for now, give additional one-time setup instructions to run that script (to manually place the jsoncpp tree in the same place the VS build puts it) and then build jsoncpp.
- Write the GN rules to just expect the built library to be in the right place.

That would mean that deciding how to pull the jsoncpp tree and implementing a GN jsoncpp build rule would become independent problems, rather than changing any one aspect having ripple effects in the other parts.

Reply:
Alrighty, could be a good time to start looking at #152 to deal with this automatically.

Reply:
Should I create `.bat` files for this or just add in the instructions the path to `dart.exe`?

Reply:
> Should I create .bat files for this or just add in the instructions the path to dart.exe?

I've been thinking about making generic run_dart_tool shell and batch scripts to avoid the duplication of the wrapper logic (finding path, ensuring cache) anyway, and this seems like a good time for it. I can do that as a quick separate PR, and then the instructions can have people use those. That will keep the instructions simple, without having to add several (hopefully short-lived) batch scripts just for this.

> And do you want me to have the script clone jsoncpp to library/windows/third_party/jsoncpp and copy the output library to library/windows/dependencies/JSON or just clone it into library/windows/dependencies/JSON

Unless it ends up being problematic let's keep the same structure that the existing batch script had, so clone to third_party then copy to dependencies. Having a clear separation between the source location and the relied-upon-by-build-systems artifacts should make things a bit easier to adjust later.